### DISCUSSION TITLE
Query - retweets and user handles

### DISCUSSION THREAD

**Discussion by claire-research** [cjbarrie/academictwitteR#280] [Q&A]
Hi there!

I'm using 'academictwitteR' in RStudio to make some queries and I've run into a couple of issues. I wanted to check in the discussion as I think it's probably user error on my part, but I can't see where I'm going wrong. 

I want to return both retweets and non-retweets, however when I tested the retweet argument (is_retweet = 'TRUE'/'FALSE'/NULL) all three datasets seemed to include all tweets. For example is_retweet = TRUE also returns non-retweets and vice versa. It seems that it is pulling the same data whatever it's set to. 

Additionally, I would like to obtain screen names for the tweets but I'm also having difficulty with this and so any guidance on how I would include the correct code would be much appreciated. 

The code I am using is: 

 tweets <-
  get_all_tweets(
    start_tweets = "2020-12-01T00:00:00Z",
    end_tweets = "2021-11-30T00:00:00Z",
    query = "(KEYWORD)",
    lang = "en",
    country = "GB",
    is_retweet = NULL,
    n = 1000,
    file = "NAME_HERE",
    data_path = "NAME_HERE"
  )

Thanks very much in advance for any help!

**Comment by chainsawriot** [comment]
@claire-research It is not pulling the "same" data. However, the `is:retweet` filter provided by Twitter is not perfect. But I recommend you checking the `sourcetweet_type` column provided by the tidy format.

``` r
require(academictwitteR)
#> Loading required package: academictwitteR

temp_path <- academictwitteR:::.gen_random_dir()
get_all_tweets("data @twitterdev", start_tweets = "2020-12-01T00:00:00Z", end_tweets = "2021-01-01T00:00:00Z", is_retweet = FALSE, data_path = temp_path, bind_tweets = FALSE, verbose = FALSE)
x <- bind_tweets(temp_path, output_format = "tidy")
x$sourcetweet_type
#> [1] NA NA NA NA NA NA NA

temp_path <- academictwitteR:::.gen_random_dir()
get_all_tweets("data @twitterdev", start_tweets = "2020-12-01T00:00:00Z", end_tweets = "2021-01-01T00:00:00Z", is_retweet = TRUE, data_path = temp_path, bind_tweets = FALSE, verbose = FALSE)
y <- bind_tweets(temp_path, output_format = "tidy")
y$sourcetweet_type
#>  [1] "retweeted" "retweeted" "retweeted" "retweeted" "retweeted" "retweeted"
#>  [7] "retweeted" "retweeted" "retweeted" "retweeted" "retweeted" "retweeted"
```

<sup>Created on 2022-02-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Reply by claire-research** [reply]
Thank you very much - I will check this out!

**Reply by claire-research** [reply]
Hello again @chainsawriot,

Firstly, thanks very much for your reply, it's been a huge help.

After having a look at this in more detail and using a couple more arguments in my query, I notice that it does work as expected for `is_retweet = FALSE` for everything that I've thrown at it. However, when adding `country = "GB"` in conjunction with `is_retweet = TRUE` it returns the error:
```
Error in `stop_subscript()`:
! Can't rename columns that don't exist.
x Column `id` doesn't exist.

```

When I run `rlang::last_trace()` it returns:
```
<error/vctrs_error_subscript_oob>
Error in `stop_subscript()`:
! Can't rename columns that don't exist.
x Column `id` doesn't exist.
Backtrace:
     ▆
  1. ├─academictwitteR::bind_tweets(temp_path, output_format = "tidy")
  2. │ └─academictwitteR:::.flat(data_path, output_format = output_format)
  3. │   └─purrr::map_dfr(data_files, convert_json, output_format = output_format)
  4. │     └─purrr::map(.x, .f, ...)
  5. │       └─academictwitteR .f(.x[[i]], ...)
  6. │         └─academictwitteR:::.gen_raw(...)
  7. │           └─dplyr::select_if(df, Negate(is.list)) %>% ...
  8. ├─dplyr::rename(., pki = tidyselect::all_of(pkicol))
  9. └─dplyr:::rename.data.frame(., pki = tidyselect::all_of(pkicol))
 10.   └─tidyselect::eval_rename(expr(c(...)), .data)
 11.     └─tidyselect:::rename_impl(...)
 12.       └─tidyselect:::eval_select_impl(...)
 13.         ├─tidyselect:::with_subscript_errors(...)
 14.         │ ├─base::tryCatch(...)
 15.         │ │ └─base tryCatchList(expr, classes, parentenv, handlers)
 16.         │ │   └─base tryCatchOne(expr, names, parentenv, handlers[[1L]])
 17.         │ │     └─base doTryCatch(return(expr), name, parentenv, handler)
 18.         │ └─tidyselect:::instrument_base_errors(expr)
 19.         │   └─base::withCallingHandlers(...)
 20.         └─tidyselect:::vars_select_eval(...)
 21.           └─tidyselect:::walk_data_tree(expr, data_mask, context_mask)
 22.             └─tidyselect:::eval_c(expr, data_mask, context_mask)
 23.               └─tidyselect:::reduce_sels(node, data_mask, context_mask, init = init)
 24.                 └─tidyselect:::walk_data_tree(new, data_mask, context_mask)
 25.                   └─tidyselect:::as_indices_sel_impl(...)
 26.                     └─tidyselect:::as_indices_impl(x, vars, strict = strict)
 27.                       └─tidyselect:::chr_as_locations(x, vars)
 28.                         └─vctrs::vec_as_location(x, n = length(vars), names = vars)
 29.                           └─vctrs `<fn>`()
 30.                             └─vctrs:::stop_subscript_oob(...)
 31.                               └─vctrs:::stop_subscript(...)
 32.                                 └─rlang::abort(...)
```
As mentioned above, this error seems to only be present when adding `is_retweet = TRUE` to the query.
Thanks so much for your time, it's greatly appreciated.

**Reply by chainsawriot** [reply]
@claire-research #290

**Reply by claire-research** [reply]
Thanks @chainsawriot. The fix in post #290 unfortunately did not work for me. However, I think the issue is that location data is not available for retweets. This explains why all location related parameters seem not to work for the retweets (e.g. country = , place = , point_radius = ), but others do. 

https://developer.twitter.com/en/docs/twitter-api/tweets/search/integrate/build-a-query#list 

'This operator will not match on Retweets, since Retweet's places are attached to the original Tweet', 

Thank you for all of your help - it is much appreciated!

**Comment by sigmasixleads** [comment]
Hello! @claire-research 

I was wondering if you were able to figure out how to get the screen names of the users? Would be appreciated if you share with me the command/code.

**Reply by claire-research** [reply]
Hi @sigmasixleads  - sorry for my delayed reply.

The code supplied in the previous message from chainsawriot worked for me - I think it's the fact that it is output using the tidy format;

y <- bind_tweets(temp_path, output_format = "tidy")

I'm very much learning all of this so if this doesn't work then it might be worth making a specific thread for this query to see if others can clarify. 

Sorry this might not be much help!
